### ISSUE TITLE
Dialyxir attempts to load transitive optional dependencies

### ISSUE THREAD

**Issue by kipcole9** [jeremyjh/dialyxir#502]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): Elixir 1.15.0-rc.0, OTP 26.0

* Which version of Dialyxir are you using? (cat mix.lock | grep dialyxir): 1.3.0

### Current behavior

Dialyxir is issuing warnings that it can't load applications that are transitive optional dependencies. An example using the [ex_cldr](https://github.com/elixir-cldr/cldr) lib:
```elixir
kip@Kips-MacBook-Pro cldr % asdf current
elixir          1.15.0-rc.0-otp-26 /Users/kip/Development/cldr/.tool-versions
erlang          26.0            /Users/kip/Development/cldr/.tool-versions
rust            stable          /Users/kip/.tool-versions
kip@Kips-MacBook-Pro cldr % mix dialyzer
Finding suitable PLTs
Checking PLT...
Error loading castore, dependency list may be incomplete.
 {~c"no such file or directory", ~c"castore.app"}
Error loading certifi, dependency list may be incomplete.
 {~c"no such file or directory", ~c"certifi.app"}
[:asn1, :castore, :certifi, :cldr_utils, :compiler, :crypto, :decimal, :earmark_parser, :eex, :elixir, :ex_doc, :ex_unit, :expo, :gettext, :inets, :jason, :kernel, :logger, :makeup, :makeup_elixir, :makeup_erlang, :mix, :nimble_parsec, :public_key, :ssl, :stdlib, :sweet_xml, :xmerl]
PLT is up to date!
ignore_warnings: .dialyzer_ignore_warnings

Starting Dialyzer
[
  check_plt: false,
  init_plt: ~c"/Users/kip/Development/cldr/_build/dev/dialyxir_erlang-26.0_elixir-1.15.0-rc.0_deps-dev.plt",
  files: [~c"/Users/kip/Development/cldr/_build/dev/lib/ex_cldr/ebin/Elixir.Cldr.LanguageTag.Chars.Cldr.LanguageTag.U.beam",
   ~c"/Users/kip/Development/cldr/_build/dev/lib/ex_cldr/ebin/Elixir.Cldr.Normalize.CalendarEra.beam",
   ~c"/Users/kip/Development/cldr/_build/dev/lib/ex_cldr/ebin/Elixir.Cldr.NoParentError.beam",
   ~c"/Users/kip/Development/cldr/_build/dev/lib/ex_cldr/ebin/Elixir.Cldr.UnknownFormatError.beam",
   ~c"/Users/kip/Development/cldr/_build/dev/lib/ex_cldr/ebin/Elixir.Mix.Tasks.Cldr.Install.Locales.beam",
   ...],
  warnings: [:unknown]
]
Total errors: 0, Skipped: 0, Unnecessary Skips: 0
done in 0m14.38s
done (passed successfully)
```
The libs `:castore` and `:certifi` are optional dependencies of [cldr_utils](https://github.com/elixir-cldr/cldr_utils). `cldr_utils` is a required dependency of `ex_cldr`.

The warning messages are unique to Elixir 1.15 / OTP 26.  

I don’t see this output for any other combinations I’ve tried:

* Elixir 1.15.0-rc.0-otp-25 with Erlang 25.3.2
* Elixir 1.14.4-otp-26 with Erlang 26.0
* Elixir 1.14.4-otp-25 with Erlang 25.3.2

Per [this comment](https://elixirforum.com/t/elixir-v1-15-0-rc-0-released/56019/27) it seems reasonable that the addition of an `optional_dependencies` config key to `.app` files in OTP 24 and now the building of that config key in Elixir 1.15 may be part of the issue.  The relevant code in dialyxir seems to be: https://github.com/jeremyjh/dialyxir/blob/e9556460bc130287dd6687ab02041634ac827913/lib/dialyxir/project.ex#L383

### Expected behavior

No warnings are issues when attempting to load transitive optional dependencies.

**Comment by jeremyjh** [comment]
@kipcole9 Thanks for the well-researched report, and thanks for your work on `cldr` & `money`. Will take a look.

**Comment by tomtaylor** [comment]
Does anyone know if there's a workaround for this at the moment?

**Comment by dvic** [comment]
> Does anyone know if there's a workaround for this at the moment?

Not sure if this helps but linking from the original discussion: https://elixirforum.com/t/elixir-v1-15-0-rc-0-released/56019/28 the suggestion made by Jose:

*When they traverse dependencies, they need to include the parent’s optional applications and not log an error if the dependency you cannot load is optional. But, from looking at the code, the warning causes no harm either.*

**Comment by carlogilmar** [comment]
Hi! After upgrade a project with Elixir `1.15.0-otp-26` with Erlang `26.0.1`, and after fix the problem with dialyzer (I had to include `prune_code_paths: false` in `mix.exs`), I got similar errors, dialyzer works, but I'm worried about all of this warnings: 

![image](https://github.com/jeremyjh/dialyxir/assets/17634377/efa46808-1db3-4f9d-be82-9508993e1f90)

**Comment by jeremyjh** [comment]
Fixed in https://github.com/jeremyjh/dialyxir/pull/514

**Comment by kipcole9** [comment]
Thank you very much @jeremyjh - and for maintaining such an important library.  Really appreciated. Even if it still does my head in from time to time!

**Comment by jeremyjh** [comment]
Published to hex in 1.14
